### PR TITLE
prefer forceFlush from provider

### PR DESCRIPTION
**Pull request by kubawach**
Improvement for opentelemetry-instrumentation-aws-lambda

## Which problem is this PR solving?

Currently the instrumentation gets `forceFlush` handler directly from the `activeSpanProcessor` of the current `TracerProvider`. The active span processor can, however, be changed during lifetime of a tracer provider. In such situation the instrumentation will call wrong `forceFlush` method. Also, a tracer provider can have multiple registered span processors and we should be flushing all of them.

## Short description of the changes

Base implementation of a TraceProvider fortunately has the `forceFlush` method implemented. Therefore, if available, we should call TP's implementation and let the provider flush all span processors.

### CODE REVIEW

**Comment by willarmiros** [comment]
> Base implementation of a TraceProvider fortunately has the forceFlush method implemented

Does the implementation just call `forceFlush` on all its span processors?

@anuraaga does this change make sense to you or would it have a risk of changing the behavior?

**Comment by kubawach** [comment]
For a broader context - current solution seems to cause bug in opentelemetry-lambda usage. If the instrumentation is initialised before proper SpanProcessor is set (which can happen due to async nature of node.js execution) the first span is sent to Noop.

**Review by anuraaga — APPROVED** [review]

**Comment by kubawach** [comment]
@willarmiros @NathanielRN any chances of looking into this? I see that @anuraaga already approved.

**Review by NathanielRN — APPROVED** [review]
This looks good to me because as @kubawach said the `BasicTracerProvider` in `opentelemetry-sdk-trace-base` has a `forceFlush` method that will call all the SpanProccesors in turn and `forceFlush` them.

See it [at this link](https://github.com/open-telemetry/opentelemetry-js/blob/5e9c7e1cb6ab57b92a59def81de2562760671c97/packages/opentelemetry-sdk-trace-base/src/BasicTracerProvider.ts#L153).

It also falls back to calling `forceFlush` on the `activeSpanProcessor` if the above fails.

@kubawach can you _add_ a test case that confirms this fall back behavior? i.e. `it('should use the activeSpanProcessor forceFlush if one is not available on the TracerProvider')`?

**Comment by kubawach** [comment]
@NathanielRN sure, will add a test for that.

File: plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts
Comment (anuraaga) [inline plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts:248]: I have a feeling that at the time of writing, there wasn't any forceFlush on the tracer provider. Now that there is, it seems like this fallback isn't actually needed anymore (and would make the code closer to what the spec is, which is that SDK TracerProvider has forceFlush)

Comment (kubawach) [reply plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts:248]: Yeah, I agree. Neither "forceFlush" nor fallback property "getActiveSpanProcessor" is in the API (TraceProvider TS type) so since we depend on the SDK implementation we can simplify things.

**Review by mhennoch — APPROVED** [review]

**Review by johnbley — APPROVED** [review]
LGTM, though it might warrant a comment on `setTracerProvider` to the extent that the param should be a `BasicTracerProvider` or otherwise implement `forceFlush` to get the desired behavior.

**Review by willarmiros — APPROVED** [review]
lgtm based on conversation

**Comment by kubawach** [comment]
@open-telemetry/javascript-maintainers any chances of getting this merged?

**Review by dyladan — APPROVED** [review]

**Comment by dyladan** [comment]
LGTM i'll merge after the test run

**Review by vmarchaud — APPROVED** [review]

**Pull request merged**
